common/components/merging.hpp.inc

```diff
@@ -0,0 +1,237 @@
+    int a_idx;
+    /** The index from b that is being merged in the current thread. */
+    int b_idx;
+    /** The number of elements from a that have been merged in total. */
+    int a_advance;
+    /** The number of elements from b that have been merged in total. */
+    int b_advance;
+};
+
+
+template <typename ValueType, typename IndexType>
+__device__ ValueType
+checked_load(const ValueType *p, IndexType i, IndexType size,
+             ValueType sentinel = device_numeric_limits<ValueType>::max)
+{
+    return i < size ? p[i] : sentinel;
+}
+
+
+}  // namespace detail
+
+
+/**
+ * @internal
+ * Warp-parallel merge algorithm that merges the first `warp_size` elements from
+ * two ranges, where each warp stores a single element from each range.
+ * It assumes that the elements are sorted in ascending order, i.e. for i < j,
+ * the value of `a` at thread i is smaller or equal to the value at thread j,
+ * and the same holds for `b`.
+ *
+ * This implementation is based on ideas from  Green et al.,
+ * "GPU merge path: a GPU merging algorithm", but uses random-access warp
+ * shuffles instead of shared-memory to exchange values of a and b.
+ *
+ * @param a      the element from the first range
+ * @param b      the element from the second range
+ * @param size   the number of elements in the output range
+ * @param group  the cooperative group that executes the merge
+ * @return  a structure containing the merge result distributed over the group.
+ */
+template <int group_size, typename ValueType, typename Group>
+__device__ detail::merge_result<ValueType> group_merge_step(ValueType a,
+                                                            ValueType b,
+                                                            Group group)
+{
+    // thread i takes care of ith element of the merged sequence
+    auto i = int(group.thread_rank());
+
+    // we want to find the smallest index `x` such that a[x] >= b[i - x - 1]
+    // or `i` if no such index exists
+    //
+    // if x = i then c[0...i - 1] = a[0...i - 1]
+    //     => merge a[i] with b[0]
+    // if x = 0 then c[0...i - 1] = b[0...i - 1]
+    //     => merge a[0] with b[i]
+    // otherwise c[0...i - 1] contains a[0...x - 1] and b[0...i - x - 1]
+    //   because the minimality of `x` implies
+    //   b[i - x] >= a[x - 1]
+    //   and a[x] >= a[0...x - 1], b[0...i - x - 1]
+    //     => merge a[x] with b[i - x]
+    auto minx = synchronous_fixed_binary_search<group_size>([&](int x) {
+        auto a_remote = group.shfl(a, x);
+        auto b_remote = group.shfl(b, max(i - x - 1, 0));
+        return a_remote >= b_remote || x >= i;
+    });
+
+    auto a_idx = minx;
+    auto b_idx = max(i - minx, 0);
+    auto a_val = group.shfl(a, a_idx);
+    auto b_val = group.shfl(b, b_idx);
+    auto cmp = a_val < b_val;
+    auto a_advance = int(popcnt(group.ballot(cmp)));
+    auto b_advance = int(group.size()) - a_advance;
```

Comment:
all involved values are <= 64, so I would not expect any issues here. Also, `group.size()` returns an unsigned value, so this conversion would fail. I could remove the cast from `popcnt` since it returns an `int` anyways, but would like to leave the rest as it is.

Reply:
If you really want to keep the current status, I will not stop you from doing so.
I simply dislike any kind of C-style casts, especially if it is easy to replace it with a C++ cast.

Reply:
For casting integer types, `static_cast<...>` looks unnecessarily verbose to me. In case you are not aware, for non-cv qualified integer types, a functional-style cast effectively is a `static_cast`, since [all other options](https://en.cppreference.com/w/cpp/language/explicit_cast) (const_cast, reinterpret_cast) do not apply there

Reply:
Yes, I know that. However, I still prefer the C++ casts for 2 reasons:
1. For consistency, so it is unlikely I ever make the mistake to do a functional cast with a template parameter or an unknown type
2. I personally like the verbosity of these casts, since it is extremely easy to spot them (especially during debugging). Everyone who is reading `static_cast` knows instantly that a cast is happening, while it is easy to over-read `int()`.

Reply:
I see, I need to think about the implications a bit. But would you consider it okay to leave this cast in?

Reply:
Yes, you can leave it in, it was just a sugesetion and I don't consider it absolutely mandatory to use C++ style casts (**in this context**).
As you said, there is pretty much next to no way that this cast can cause any issues, so you can leave it in.